### PR TITLE
Azure 1 june

### PR DESCRIPTION
**Pull request by tbartynski**

### CODE REVIEW

File: app/services/atmosphere/cloud/vm_creator.rb [hunk 1]
```diff
@@ -27,7 +20,26 @@ def execute
...
+      case (@tmpl.compute_site.technology)
+      when 'azure'
+        server_params[:vm_name] = SecureRandom.uuid
+        server_params[:vm_user] = Atmosphere.azure_vm_user
+        server_params[:password] = Atmosphere.azure_vm_password
+        server_params[:image] = tmpl_id
+        server_params[:location] = 'North Central US'
+        server_params[:vm_size] = flavor_id
+        set_azure_endpoints(server_params)
+      else # TODO: Untangle AWS and OpenStack
```
Comment (mkasztelnik) [inline app/services/atmosphere/cloud/vm_creator.rb]: This TODO need to be handled before PR merge?

File: app/services/atmosphere/cloud/vm_creator.rb [hunk 2]
```diff
@@ -27,7 +20,26 @@ def execute
         server_params[:nics] = [{ net_id: @nic }]
       end
 
-      set_security_groups!(server_params)
+      case (@tmpl.compute_site.technology)
+      when 'azure'
+        server_params[:vm_name] = SecureRandom.uuid
+        server_params[:vm_user] = Atmosphere.azure_vm_user
+        server_params[:password] = Atmosphere.azure_vm_password
```
Comment (mkasztelnik) [inline app/services/atmosphere/cloud/vm_creator.rb]: This is quite tricky...

File: app/services/atmosphere/cloud/vm_creator.rb [hunk 3]
```diff
@@ -39,6 +51,30 @@ def execute
...
+            udp_endpoints_str << "#{pmt.target_port}:#{pmt.target_port},"
+          end
+        end
+      if tcp_endpoints_str.present?
+        params[:tcp_endpoints] = tcp_endpoints_str.chomp!(',')
+      end
+      if udp_endpoints_str.present?
+        params[:udp_endpoints] = udp_endpoints_str.chomp!(',')
+      end
+    end
```
Comment (mkasztelnik) [inline app/services/atmosphere/cloud/vm_creator.rb]: I understand that only predefined PMT are supported. There will not be possible to add new redirection once appliance is started.

**Comment by mkasztelnik** [comment]
Please squash it into one commit (to remove it in the feature easily :smile:).

File: app/services/atmosphere/cloud/vm_creator.rb (hunk 1 above)
Comment (tbartynski) [inline app/services/atmosphere/cloud/vm_creator.rb]: Nope

File: app/services/atmosphere/cloud/vm_creator.rb (hunk 2 above)
Comment (tbartynski) [inline app/services/atmosphere/cloud/vm_creator.rb]: Hound has had a similar comment. Suggest a better solution. I'm the author of this part so I'm very subjective about understanding this code.

File: app/services/atmosphere/cloud/vm_creator.rb (hunk 3 above)
Comment (tbartynski) [inline app/services/atmosphere/cloud/vm_creator.rb]: We do not support dev mode at all since saving tempates is not supported.

**Comment by tbartynski** [comment]
@mkasztelnik Isn't a merge of a branch another commit that we could easily remove in future?

File: app/services/atmosphere/cloud/vm_creator.rb (hunk 2 above)
Comment (mkasztelnik) [inline app/services/atmosphere/cloud/vm_creator.rb]: The problem is that all VMs started on Azure will have the same username/password, unfortunately I don't have a better solution without adding new things into atmosphere :cry:

File: app/services/atmosphere/cloud/vm_creator.rb (hunk 3 above)
Comment (mkasztelnik) [inline app/services/atmosphere/cloud/vm_creator.rb]: Ok

**Comment by mkasztelnik** [comment]
@tbartynski not really, because finding this merge request commit is sometime hard. I will insist to squash it into one commit, since most probably we will remove  it from the atmosphere or extract into separate gem in the near feature.

To squash into one commit use

```
git rebase -i HEAD~number_of_commits
```

See http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html for details.

File: lib/generators/templates/atmosphere.rb [hunk 4]
```diff
@@ -117,4 +117,12 @@
   # this new class name as `ability_class` parameter value.
   #
   # config.ability_class = 'Atmosphere::Ability'
+
+  # Password for vm instantiated at Azure (ssh keys are not suported)
+  # By default azure_vm_password should be defined in air.yml
+  # config.azure_vm_password = Settings.azure_vm_password
```
Comment (nowakowski) [inline lib/generators/templates/atmosphere.rb]: Could you add a suitable placeholder to air.yml.example?

Comment (mkasztelnik) [inline lib/generators/templates/atmosphere.rb]: @nowakowski we don't have `air.yml.example` in atmosphere (it is vphshare specific). What is more we don't have `Settings` class in atmosphere. Please add here some example text.

**Comment by nowakowski** [comment]
:+1: notwithstanding placeholders in config files.

File: app/services/atmosphere/cloud/vm_creator.rb (hunk 2 above)
Comment (tbartynski) [inline app/services/atmosphere/cloud/vm_creator.rb]: I am aware of this but I could not think of anything better that would not introduce a lot of work.

File: lib/generators/templates/atmosphere.rb (hunk 4 above)
Comment (tbartynski) [inline lib/generators/templates/atmosphere.rb]: Done in ce63407.

**Comment by tbartynski** [comment]
I tried an interactive rebase but it is very complex and error prone. I had to resolve plenitude of conflicts and quit doing that after 4 hours.
I agreed with @mkasztelnik to create a diff between this brnach and master and then manually purge any azure-specific code from master.

**Pull request merged**